checker-rs/src/main.rs

```diff
@@ -0,0 +1,327 @@
+
+        client.register(&user).await?;
+        client.login(user).await?;
+
+        client.set_note("sjgorisdr").await?;
+
+        self.store_user(client.user.unwrap()).await
+    }
+
+    async fn check_noise_note(&self, checker_request: &CheckerRequest) -> CheckerResult<()> {
+        let user = self
+            .find_user(&checker_request.task_chain_id)
+            .await?
+            .ok_or(CheckerError::Mumble("Could not find old user"))?;
+        let mut client = NotebookClient::connect(checker_request).await?;
+
+        client.login(user).await?;
+        let noise = client.get_note().await?;
+
+        if &noise != client.user.as_ref().unwrap().note.as_ref().unwrap() {
+            warn!(
+                "Expected noise note with content {}, instead got {}",
+                client.user.unwrap().note.unwrap(),
+                noise
+            );
+            return Err(CheckerError::Mumble("Invalid note retrieved"));
+        }
+        Ok(())
+    }
+
+    async fn havoc_help(&self, checker_request: &CheckerRequest) -> CheckerResult<()> {
+        // Well this will mean that not all service-details will get test-coverage,
+        // however both lines are way too similar to warrant separete variants
+        let help_text = if random() {
+            info!("Getting help in an authenticated state");
+            let user = NotebookUser::gen_random(checker_request.task_chain_id.clone());
+            let mut client = NotebookClient::connect(checker_request).await?;
+            client.register(&user).await?;
+            client.login(user).await?;
+            client.get_help().await?
+        } else {
+            info!("Getting help immediately");
+            let mut client = NotebookClient::connect(checker_request).await?;
+            client.get_help().await?
+        };
+
+        if help_text.as_str() != HELP_TEXT {
+            return Err(CheckerError::Mumble("Invalid Help"));
+        }
+
+        Ok(())
+    }
+
+    async fn havoc_user(&self, checker_request: &CheckerRequest) -> CheckerResult<()> {
+        info!("Getting userlist immediately");
+        let user = NotebookUser::gen_random(checker_request.task_chain_id.clone());
+        let mut client = NotebookClient::connect(checker_request).await?;
+        client.register(&user).await?;
+
+        let user1 = user.clone();
+        let request2 = (*checker_request).clone();
+
+        // Test the userlist as the user itself
+        let future_auth = async move {
+            info!("Foo");
+            client.login(user1).await?;
+            client.get_users().await
+        };
+
+        // And as any other connected user
+        let future_unauth = async move {
+            info!("Foo");
+            let mut client = NotebookClient::connect(&request2).await?;
+            client.get_users().await
+        };
+
+        let user_lists: (CheckerResult<_>, CheckerResult<_>) = try_join!(
```

Comment:
Spawn + try_join is able to run both requests in parralel since spawn `may` move tasks onto different threads of the executor. It may not be as obvious as running the tasks normally, but I wanted to show how to await multiple futures simultaneously in rust.

`tracing` is a fancy (and quite powerful) logging framework, from the developers of tokio.
By "instrumenting" both futures into different spans, log-messages from both tasks will be tagged slightly differently.

Reply:
Maybe add en explanation for that ;)